codalab/lib/beam/MultiReaderFileStream.py

```diff
@@ -25,33 +26,60 @@ def read(s, num_bytes=None):
 
             def peek(s, num_bytes):
                 return self.peek(s._index, num_bytes)
+            
+            def seek(s, offset, whence=SEEK_SET):
+                return self.seek(s._index, offset, whence)
 
         self.readers = [FileStreamReader(i) for i in range(0, self.NUM_READERS)]
 
-    def _fill_buf_bytes(self, index: int, num_bytes=None):
-        with self._lock:
-            while num_bytes is None or len(self._bufs[index]) < num_bytes:
-                s = self._fileobj.read(num_bytes)
-                if not s:
-                    break
-                for i in range(0, self.NUM_READERS):
-                    self._bufs[i].write(s)
-
-    def read(self, index: int, num_bytes=None):  # type: ignore
+    def _fill_buf_bytes(self, num_bytes=None):
+        # with self._lock:
+        s = self._fileobj.read(num_bytes)
+        if not s:
+            return
+        self._buffer += s
+        self._size += len(s)
+
+    def read(self, index: int, num_bytes=0):  # type: ignore
         """Read the specified number of bytes from the associated file.
         index: index that specifies which reader is reading.
         """
-        self._fill_buf_bytes(index, num_bytes)
-        if num_bytes is None:
-            num_bytes = len(self._bufs[index])
-        s = self._bufs[index].read(num_bytes)
-        self._pos[index] += len(s)
+        # Calculate how many new bytes need to be read
+        with self._lock:
+            new_bytes_needed = num_bytes - (max(self._pos) - self._pos[index])
```

Comment:
I think it's clearer to define a new variable for `new_pos = self._pos[index] + num_bytes`, which can be re-used...

Reply:
I'm a little bit unclear what this would be used for?

Reply:
Ahh ok I see, that does make it clearer

Reply:
Made the changes